gobblin-data-management/src/main/java/gobblin/data/management/copy/TimestampBasedCopyableDataset.java

```diff
@@ -0,0 +1,227 @@
+import gobblin.data.management.policy.VersionSelectionPolicy;
+import gobblin.data.management.version.TimestampedDatasetVersion;
+import gobblin.data.management.version.finder.DateTimeDatasetVersionFinder;
+import gobblin.data.management.version.finder.VersionFinder;
+import gobblin.dataset.FileSystemDataset;
+import gobblin.util.filters.HiddenFilter;
+import gobblin.util.ExecutorsUtils;
+import gobblin.util.PathUtils;
+import gobblin.util.executors.ScalingThreadPoolExecutor;
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+
+
+/**
+ * Implementation of {@link CopyableDataset}. It depends on {@link #versionSelectionPolicy} to select the
+ * dataset versions for copying.
+ *
+ * The default logic ({@link #isCopyableFile(Path, FileSystem, FileSystem)}) for determining if a file is
+ * {@link CopyableFile} is based on the file existence and modified_timestamp at source and target {@link
+ * FileSystem}s.
+ */
+@Slf4j
+@Getter
+@SuppressWarnings("unchecked")
+public class TimestampBasedCopyableDataset implements CopyableDataset, FileSystemDataset {
+
+  private final Path datasetRoot;
+  private final String pattern;
+  private final DateTimeZone dateTimeZone;
+  private final VersionFinder<TimestampedDatasetVersion> datasetVersionFinder;
+  private final VersionSelectionPolicy<TimestampedDatasetVersion> versionSelectionPolicy;
+  private final ExecutorService executor;
+  private final FileSystem srcFs;
+
+  public static final String DATASET_VERSION_FINDER = "timestamp.based.copyable.dataset.version.finder";
+  public static final String DEFAULT_DATASET_VERSION_FINDER = DateTimeDatasetVersionFinder.class.getName();
+
+  public static final String COPY_POLICY = "timestamp.based.copyable.dataset.copy.policy";
+  public static final String DEFAULT_COPY_POLICY = TimeBasedSelectionPolicy.class.getName();
+
+  public static final String THREADPOOL_SIZE_TO_GET_COPYABLE_FILES = "threadpool.size.to.get.copyable.files";
+  public static final String DEFAULT_THREADPOOL_SIZE_TO_GET_COPYABLE_FILES = "20";
+
+  public TimestampBasedCopyableDataset(FileSystem fs, Properties props, Path datasetRoot) throws IOException {
+    this.srcFs = fs;
+    this.datasetRoot = datasetRoot;
+    this.pattern = props.getProperty(DateTimeDatasetVersionFinder.DATE_TIME_PATTERN_KEY);
+    this.dateTimeZone =
+        DateTimeZone.forID(props.getProperty(DateTimeDatasetVersionFinder.DATE_TIME_PATTERN_TIMEZONE_KEY,
+            gobblin.data.management.version.finder.DateTimeDatasetVersionFinder.DEFAULT_DATE_TIME_PATTERN_TIMEZONE));
+
+    try {
+      Class<?> copyPolicyClass = Class.forName(props.getProperty(COPY_POLICY, DEFAULT_COPY_POLICY));
+      this.versionSelectionPolicy =
+          (VersionSelectionPolicy<TimestampedDatasetVersion>) copyPolicyClass.getConstructor(Properties.class)
+              .newInstance(props);
+      Class<?> timestampedDatasetVersionFinderClass =
```

Comment:
Are other version finder classes really supported? What are the restrictions? Is it any class that implements VersionFinder<TimestampedDatasetVersion> or should it be a subclass DateTimeDatasetVersionFinder? The code has a dependency on some config properties for DateTimeDatasetVersionFinder. What will happen if we inject some completely different class?

I think it is OK if the version finder is fixed unless you see specific use-cases to be replaced. 

Further, the current implementation re-uses some configuration properties from DateTimeDatasetVersionFinder but it seems like the logic to read those is slightly different (e.g. for DATE_TIME_PATTERN_TIMEZONE_KEY). Can't we just depend on the DateTimeDatasetVersionFinder instance to get those?